src/util/enriched_string.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright (C) 2013 xyz, Ilya Zhuravlev <whatever@xyz.is>
+Copyright (C) 2016 Nore, Nathanaël Courant <nore@mesecons.net>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "enriched_string.h"
+#include "util/string.h"
+#include "log.h"
+EnrichedString::EnrichedString()
+{
+	clear();
+}
+
+EnrichedString::EnrichedString(const std::wstring &string, const std::vector<SColor> &colors):
+	m_string(string),
+	m_colors(colors),
+	m_has_background(false)
+{}
+
+EnrichedString::EnrichedString(const std::wstring &s, const SColor color)
+{
+	clear();
+	addAtEnd(s, color);
+}
+
+EnrichedString::EnrichedString(const wchar_t *str, const SColor color)
+{
+	clear();
+	addAtEnd(std::wstring(str), color);
+}
+
+void EnrichedString::operator=(const wchar_t *str)
+{
+	clear();
+	addAtEnd(std::wstring(str), SColor(255, 255, 255, 255));
+}
+
+void EnrichedString::addAtEnd(const std::wstring &s, const SColor initial_color)
+{
+	SColor color(initial_color);
+	size_t i = 0;
+	while (i < s.length()) {
+		if (s[i] == L'\x1b') {
```

Comment:
Since this if statement body is the majority of the function, wouldn't it be better to invert the condition so that the non-color-code case continues early?  You could save a level of indentation that way.